dpnp/dpnp_algo/dpnp_fill.py

```diff
@@ -0,0 +1,85 @@
+# -*- coding: utf-8 -*-
+# *****************************************************************************
+# Copyright (c) 2016-2024, Intel Corporation
+# All rights reserved.
+#
+# Redistribution and use in source and binary forms, with or without
+# modification, are permitted provided that the following conditions are met:
+# - Redistributions of source code must retain the above copyright notice,
+#   this list of conditions and the following disclaimer.
+# - Redistributions in binary form must reproduce the above copyright notice,
+#   this list of conditions and the following disclaimer in the documentation
+#   and/or other materials provided with the distribution.
+#
+# THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS "AS IS"
+# AND ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE
+# IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE
+# ARE DISCLAIMED. IN NO EVENT SHALL THE COPYRIGHT HOLDER OR CONTRIBUTORS BE
+# LIABLE FOR ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR
+# CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF
+# SUBSTITUTE GOODS OR SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS
+# INTERRUPTION) HOWEVER CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN
+# CONTRACT, STRICT LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE)
+# ARISING IN ANY WAY OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF
+# THE POSSIBILITY OF SUCH DAMAGE.
+# *****************************************************************************
+
+import dpctl.tensor as dpt
+import numpy as np
+from dpctl.tensor._tensor_impl import (
+    _copy_usm_ndarray_into_usm_ndarray,
+    _full_usm_ndarray,
+    _zeros_usm_ndarray,
+)
+from dpctl.utils import SequentialOrderManager
+
+import dpnp
+from dpnp.dpnp_array import dpnp_array
+
+
+def dpnp_fill(arr, val):
+    dpnp.check_supported_arrays_type(arr)
+    arr = dpnp.get_usm_ndarray(arr)
+    exec_q = arr.sycl_queue
+
+    dpnp.check_supported_arrays_type(val, scalar_type=True, all_scalars=True)
+    # if val is an array, process it
+    if isinstance(val, (dpnp_array, dpt.usm_ndarray)):
+        val = dpnp.get_usm_ndarray(val)
+        if val.shape != ():
+            raise ValueError("`val` must be a scalar")
+        # asarray moves scalar to the correct device
+        # and casts to the expected dtype
+        a_val = dpt.asarray(
```

Comment:
Good point, my logic was that this case is allowed for both dpnp and dpctl in `fill`, but in retrospect, that function allows it because of `device` keyword.

I'll make a change to to disallow this case.